src/Foundatio.Aliyun/Storage/AliyunFileStorage.cs

```diff
@@ -0,0 +1,199 @@
+using System.Threading.Tasks;
+using Aliyun.OSS;
+using Foundatio.Extensions;
+
+namespace Foundatio.Storage {
+    public class AliyunFileStorage : IFileStorage {
+        private readonly string _bucketName;
+        private readonly OssClient _client;
+
+        public AliyunFileStorage(string connectionString, string bucketName = "storage") {
+            var account = AliyunConnectionString.Parse(connectionString);
+            _client = account.CreateClient();
+            _bucketName = bucketName;
+            if (!_client.DoesBucketExist(_bucketName)) _client.CreateBucket(_bucketName);
+        }
+
+        public void Dispose() { }
+
+        private bool IsNotFoundException(Exception ex) {
+            if (ex is AggregateException aggregateException) {
+                foreach (var innerException in aggregateException.InnerExceptions) {
+                    if (IsNotFoundException(innerException)) {
+                        return true;
+                    }
+                }
+            }
+            if (ex is WebException webException && webException.Response is HttpWebResponse response) {
+                return response.StatusCode == HttpStatusCode.NotFound;
+            }
+            return false;
+        }
+
+        private string NormalizePath(string path) {
+            return path?.Replace('\\', '/');
+        }
+
+        public async Task<Stream> GetFileStreamAsync(string path, CancellationToken cancellationToken = default(CancellationToken)) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            var response = await Task.Factory.FromAsync(
+                (request, callback, state) => _client.BeginGetObject(request, callback, state),
+                result => _client.EndGetObject(result),
+                new GetObjectRequest(_bucketName, NormalizePath(path)), null);
+            return response.Content;
+        }
+
+        public Task<FileSpec> GetFileInfoAsync(string path) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            path = NormalizePath(path);
+            try {
+                var metadata = _client.GetObjectMetadata(_bucketName, path);
+                return Task.FromResult(new FileSpec {
+                    Path = path,
+                    Size = metadata.ContentLength,
+                    Created = metadata.LastModified,
+                    Modified = metadata.LastModified
+                });
+            }
+            catch (Exception) {
+                return Task.FromResult((FileSpec)null);
+            }
+        }
+
+        public Task<bool> ExistsAsync(string path) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            try {
+                return Task.FromResult(_client.DoesObjectExist(_bucketName, NormalizePath(path)));

```

Comment:
From reading the code it looks like if it doesn't exist they throw an web exception? What is the point of their api that returns a boolean?? I know at least in the azure client we don't try catch what they return (https://github.com/FoundatioFx/Foundatio.AzureStorage/blob/feature/logging/src/Foundatio.AzureStorage/Storage/AzureFileStorage.cs#L49) but if they are throwing a web exception in a sync method then we have to here.

Reply:
Actually the official SDK does not support netstandard, so I have to reference an unofficially netstandard ported library. But this ported library does not implemented and tested well for some points. This is an example about it. In the official SDK it can be done correctly and return false when the file does not exists, but the unofficial ported library cannot, so it will never return false. So we have to catch aggregate exception to analyze the failure scenario.